Sources/SourceKitLSP/Swift/OnDiskDocumentManager.swift

```diff
@@ -0,0 +1,169 @@
+/// open in the editor. This allows for retrieving multiple symbol graphs from the same file without having
+/// to re-open and parse the syntax tree every time.
+actor OnDiskDocumentManager: Sendable {
+  private weak var sourceKitLSPServer: SourceKitLSPServer?
+  private var openSnapshots: [DocumentURI: (snapshot: DocumentSnapshot, patchedCompileCommand: SwiftCompileCommand?)]
+
+  init(sourceKitLSPServer: SourceKitLSPServer) {
+    self.sourceKitLSPServer = sourceKitLSPServer
+    self.openSnapshots = [:]
+  }
+
+  /// Retrieve the symbol graph at a particular ``SymbolLocation``.
+  ///
+  /// A unique dummy document will be opened in sourcekitd to fulfill the request.
+  ///
+  /// - Parameter symbolLocation: The location of the symbol to fetch the symbol graph for.
+  func fetchSymbolGraph(at symbolLocation: SymbolLocation) async throws -> String? {
+    let (snapshot, patchedCompileCommand) = try await openDocumentSnapshot(at: symbolLocation)
+    let swiftLanguageService = try await swiftLanguageService(for: symbolLocation.documentUri)
+    return try await swiftLanguageService.cursorInfo(
+      snapshot,
+      compileCommand: patchedCompileCommand,
+      Range(snapshot.position(of: symbolLocation)),
+      includeSymbolGraph: true
+    ).symbolGraph
+  }
+
+  /// Open a unique dummy document in sourcekitd that has the contents of the file on disk for uri, but an arbitrary
+  /// URI which doesn't exist on disk.
+  ///
+  /// The document will be retained until ``closeAllDocuments()`` is called. This will avoid parsing the same
+  /// document multiple times if more than one symbol needs to be looked up.
+  ///
+  /// - Parameter symbolLocation: The location of the document that will be opened.
+  func openDocumentSnapshot(
+    at symbolLocation: SymbolLocation
+  ) async throws -> (snapshot: DocumentSnapshot, patchedCompileCommand: SwiftCompileCommand?) {
+    try await openDocumentSnapshot(for: symbolLocation.documentUri)
+  }
+
+  /// Open a unique dummy document in sourcekitd that has the contents of the file on disk for uri, but an arbitrary
+  /// URI which doesn't exist on disk.
+  ///
+  /// The document will be retained until ``closeAllDocuments()`` is called. This will avoid parsing the same
+  /// document multiple times if more than one symbol needs to be looked up.
+  ///
+  /// - Parameter uri: The ``DocumentURI`` that will be opened.
+  func openDocumentSnapshot(
+    for uri: DocumentURI
+  ) async throws -> (snapshot: DocumentSnapshot, patchedCompileCommand: SwiftCompileCommand?) {
+    guard let cachedSnapshot = openSnapshots[uri] else {
+      let languageService = try await swiftLanguageService(for: uri)
+      let snapshot = try await languageService.openSnapshotFromDiskOpenedInSourcekitd(
+        uri: uri,
+        fallbackSettingsAfterTimeout: false
+      )
+      openSnapshots[uri] = snapshot
+      return snapshot
+    }
+    return cachedSnapshot
```

Comment:
I would flip this to an `if` so that the majority of the code is not indented.

```suggestion
    if let cachedSnapshot = openSnapshots[uri] else {
      return cachedSnapshot
    }
    let languageService = try await swiftLanguageService(for: uri)
    let snapshot = try await languageService.openSnapshotFromDiskOpenedInSourcekitd(
      uri: uri,
      fallbackSettingsAfterTimeout: false
    )
    openSnapshots[uri] = snapshot
    return snapshot
```

Reply:
Done.